src/photos/lib/onPhotoTrashed.js

```diff
@@ -0,0 +1,52 @@
+import { getReferencedBy } from 'cozy-client'
+import log from 'cozy-logger'
+
+import { buildPhotosTrashed } from 'photos/queries/queries'
+import { DOCTYPE_ALBUMS, DOCTYPE_FILES } from 'drive/lib/doctypes'
+
+const onPhotoTrashed = async client => {
+  log('info', `Service called with COZY_URL: ${process.env.COZY_URL}`)
+
+  const photosTrashedQuery = buildPhotosTrashed()
+  try {
+    const photos = await client.queryAll(
+      photosTrashedQuery.definition,
+      photosTrashedQuery.options
+    )
+
+    if (photos.length > 0) {
+      log(
+        'info',
+        `Start deleting album references on ${photos.length} trashed photos`
+      )
+      photos.forEach(async photo => {
```

Comment:
Be careful, in that case, forEach will be call in //. So we can make a lot of requests to the stack and in that case, we open a lot of handlers (handler can be : file opening or... network request). 

We have a very low limit on the number of open handlers in prod to avoid crazy memory consumption.

So you should do that differently? 

Or do I've a bad understanding of how forEach with async works?

Reply:
Yep or p-limit (https://www.npmjs.com/package/p-limit we use it in a few konnectors i think)

Reply:
In fact, to be more precise: 
- forEach is way more performant because we do actions in // 
- `for of` do actions sync 

So in fact, forEach will be more performant in a lof of cases. Except when you have too much work in //. You can fully fill the RAM or... you can reach the opened handlers limit. 

My favorite solution for that, is to keep async, but with concurrency. With something like 20 async actions, it's ok. Like that, we still do 20 actions in //, so we have good perf, without killing the servers. 

(in our case, you make call from nsjail to the stack, no "user" network concerned or else, so the gain is not as good as when running on the client side). 

So if you want to use p-limit go for it, if you want to keep that way, go for it too ;). Your call

Reply:
Thanks for the additional explanation. I think I'll stick with `for...of` because I don't want to add a new dependency to the project and I think the syntax is pretty clean